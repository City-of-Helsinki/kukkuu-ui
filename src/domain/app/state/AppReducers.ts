import { combineReducers } from '@reduxjs/toolkit';

import authenticationReducers from '../../auth/state/AuthenticationReducers';
import profileReducer from '../../profile/state/ProfileReducers';
import registrationReducer from '../../registration/state/RegistrationReducers';
import uiReducers from './ui/UIReducers';
import EventReducers from '../../event/state/EventReducers';

export default combineReducers({
  authentication: authenticationReducers,
  event: EventReducers,
  registration: registrationReducer,
  profile: profileReducer,
  ui: uiReducers,
});
