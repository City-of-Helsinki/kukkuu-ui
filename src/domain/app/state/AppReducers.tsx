import { combineReducers } from 'redux-starter-kit';

import registrationReducer from '../../registration/state/RegistrationReducers';
export default combineReducers({ registration: registrationReducer });
