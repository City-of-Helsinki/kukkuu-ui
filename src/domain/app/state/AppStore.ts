import { configureStore, getDefaultMiddleware, Store } from 'redux-starter-kit';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import rootReducer from './AppReducers';

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['registration'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store: Store = configureStore({
  devTools: true,
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
  reducer: persistedReducer,
});

const persistor = persistStore(store);

export { persistor, store };
