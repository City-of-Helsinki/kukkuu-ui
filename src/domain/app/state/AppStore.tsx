import { createStore } from 'redux';
import { devToolsEnhancer } from 'redux-devtools-extension';
import { loadUser } from 'redux-oidc';

import userManager from '../../auth/userManager';
import rootReducer from './AppReducers';
const store = createStore(rootReducer, devToolsEnhancer({}));

loadUser(store, userManager);

export default store;
