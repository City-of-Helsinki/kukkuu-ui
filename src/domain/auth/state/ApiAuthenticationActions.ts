import { createAction } from 'redux-starter-kit';

import { API_AUTHENTICATION_ACTIONS } from '../constants/ApiAuthenticationActionConstants';

const startFetchingToken = createAction(
  API_AUTHENTICATION_ACTIONS.START_FETCHING_TOKEN
);

const fetchTokenSuccess = createAction(
  API_AUTHENTICATION_ACTIONS.FETCH_TOKEN_SUCCESS
);

const fetchTokenError = createAction(
  API_AUTHENTICATION_ACTIONS.FETCH_TOKEN_ERROR
);

export { startFetchingToken, fetchTokenSuccess, fetchTokenError };
