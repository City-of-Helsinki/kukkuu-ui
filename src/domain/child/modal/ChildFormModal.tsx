import * as React from 'react';
import { Formik, FieldArray, FormikErrors } from 'formik';
import { useTranslation } from 'react-i18next';

import Modal from '../../../common/components/modal/Modal';
import styles from './childFormModal.module.scss';
import BirthdateFormField from '../../home/form/partial/BirthdateFormField';
import EnhancedInputField from '../../../common/components/form/fields/input/EnhancedInputField';
import InputField from '../../../common/components/form/fields/input/InputField';
import Button from '../../../common/components/button/Button';
import SelectField from '../../../common/components/form/fields/select/SelectField';
import { Child } from '../types/ChildTypes';
import { getTranslatedRelationshipOptions } from '../ChildUtils';
import NavigationPropmt from '../../../common/components/prompt/NavigationPrompt';
import {
  validatePostalCode,
  validateDate,
} from '../../../common/components/form/validationUtils';
import { formatTime, newMoment } from '../../../common/time/utils';
import { BACKEND_DATE_FORMAT } from '../../../common/time/TimeConstants';
import { isChildEligible } from '../../registration/notEligible/NotEligibleUtils';
import Icon from '../../../common/components/icon/Icon';
import personIcon from '../../../assets/icons/svg/adultFace.svg';

export interface ChildFormModalValues extends Omit<Child, 'birthdate'> {
  birthdate: {
    day: number | string;
    month: number | string;
    year: number | string;
  };
  childBirthdate?: string;
}
interface ChildFormModalProps {
  initialValues: ChildFormModalValues;
  label: string;
  onSubmit: (payload: Child) => void;
  onDelete?: () => void;
  isOpen: boolean;
  setIsOpen: (value: boolean) => void;
  formType?: CHILD_FORM_TYPES;
}

export enum CHILD_FORM_TYPES {
  ADD = 'ADD',
  EDIT = 'EDIT',
}

const ChildFormModal: React.FunctionComponent<ChildFormModalProps> = ({
  initialValues,
  label,
  onSubmit,
  onDelete,
  isOpen,
  setIsOpen,
  formType = CHILD_FORM_TYPES.ADD,
}) => {
  const { t } = useTranslation();
  const [isFilling, setFormIsFilling] = React.useState(false);
  const [nonEligible, toggleNonEligible] = React.useState(false);
  const isEditForm = formType === CHILD_FORM_TYPES.EDIT;

  // Child who already have relationship can not go back to have empty relationship anymore
  // Why ? Ask backend guys.
  const isChildHavingRelationship = !!initialValues.relationship?.type;

  return (
    <div className={styles.childFormModalWrapper}>
      {isOpen && (
        <NavigationPropmt
          warningMessage={t('common.form.leave.warning.text')}
          isHalfFilling={isFilling}
        />
      )}
      <Modal
        isOpen={isOpen}
        label={nonEligible ? '' : label}
        toggleModal={(value: boolean) => {
          toggleNonEligible(false);
          setIsOpen(value);
        }}
        showLabelIcon={!nonEligible}
        setFormIsFilling={setFormIsFilling}
      >
        {nonEligible ? (
          <div className={styles.notEligible}>
            <h3>{t('registration.notEligible.title')}</h3>
            <p>{t('registration.notEligible.text')}</p>
            <Icon className={styles.icon} src={personIcon} />
            <Button
              className={styles.goBackButton}
              onClick={() => setIsOpen(false)}
            >
              {t('child.form.modal.notEligible.return.text')}
            </Button>
          </div>
        ) : (
          <Formik
            validate={(values: ChildFormModalValues) => {
              const {
                birthdate: { day, month, year },
              } = values;
              if (!isFilling) {
                setFormIsFilling(true);
              }

              const errors: FormikErrors<ChildFormModalValues> = {};

              if (day && month && year) {
                errors.childBirthdate = validateDate(`${day}.${month}.${year}`);

                if (!errors.childBirthdate) {
                  // Delete the property manually so form will be valid when this is undefined.
                  delete errors.childBirthdate;
                }
              }
              return errors;
            }}
            initialValues={initialValues}
            onSubmit={(values: ChildFormModalValues) => {
              setFormIsFilling(false);
              const child: Child = Object.assign({}, values, {
                birthdate: formatTime(
                  newMoment(
                    `${values.birthdate.year}-${values.birthdate.month}-${values.birthdate.day}`,
                    BACKEND_DATE_FORMAT
                  )
                ),
              });

              const isEligible = isChildEligible(child);
              if (isEligible) {
                onSubmit(child);
              } else {
                toggleNonEligible(true);
              }
            }}
          >
            {({ isSubmitting, handleSubmit }) => (
              <form onSubmit={handleSubmit} id="childModalForm">
                <FieldArray
                  name="birthdate"
                  render={props => <BirthdateFormField {...props} />}
                />

                <div className={styles.childInfo}>
                  <EnhancedInputField
                    className={styles.childHomeCity}
                    id="homeCity"
                    name="homeCity"
                    label={t(
                      'homePage.preliminaryForm.childHomeCity.input.label'
                    )}
                    required={true}
                    component={InputField}
                    placeholder={t(
                      'homePage.preliminaryForm.childHomeCity.input.placeholder'
                    )}
                  />

                  <EnhancedInputField
                    className={styles.childPostalCode}
                    required={true}
                    id="postalCode"
                    name="postalCode"
                    validate={validatePostalCode}
                    label={t('registration.form.child.postalCode.input.label')}
                    component={InputField}
                    placeholder={t(
                      'registration.form.child.postalCode.input.placeholder'
                    )}
                  />
                </div>

                <div className={styles.childName}>
                  <EnhancedInputField
                    id="firstName"
                    name="firstName"
                    label={t('registration.form.child.firstName.input.label')}
                    component={InputField}
                    autoComplete="new-password"
                    placeholder={t(
                      'registration.form.child.firstName.input.placeholder'
                    )}
                  />
                  <EnhancedInputField
                    id="lastName"
                    name="lastName"
                    autoComplete="new-password"
                    label={t('registration.form.child.lastName.input.label')}
                    component={InputField}
                    placeholder={t(
                      'registration.form.child.lastName.input.placeholder'
                    )}
                  />
                </div>

                <EnhancedInputField
                  id="relationship.type"
                  name="relationship.type"
                  label={t('registration.form.child.relationship.input.label')}
                  autoSelect={isChildHavingRelationship}
                  component={SelectField}
                  options={getTranslatedRelationshipOptions(t)}
                  placeholder={t(
                    'registration.form.child.relationship.input.placeholder'
                  )}
                />

                <Button
                  type="submit"
                  className={styles.submitButton}
                  disabled={isSubmitting}
                >
                  {t('child.form.modal.add.label')}
                </Button>

                {isEditForm && (
                  <Button
                    className={styles.deleteChild}
                    ariaLabel={t('child.form.modal.delete.text')}
                    onClick={() => onDelete && onDelete()}
                  >
                    {t('child.form.modal.delete.text')}
                  </Button>
                )}
              </form>
            )}
          </Formik>
        )}
      </Modal>
    </div>
  );
};

export default ChildFormModal;
