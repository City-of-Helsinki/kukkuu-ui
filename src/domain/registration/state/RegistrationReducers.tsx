import { createReducer } from 'redux-starter-kit';

import { RegistrationData } from '../types/RegistrationTypes';
import { REGISTRATION_ACTIONS } from '../constants/RegistrationActionConstants';

export const defaultRegistrationData: RegistrationData = {
  formValues: {
    child: {
      birthday: '',
      firstName: '',
      homeCity: '',
      lastName: '',
    },
    guardian: {
      firstName: '',
      lastName: '',
      phoneNumber: '',
      relationship: '',
    },
    verifyInformation: false,
  },
};

export default createReducer(defaultRegistrationData, {
  [REGISTRATION_ACTIONS.SET_FORM_VALUES]: (state, action) =>
    Object.assign({}, state, { formValues: action.payload }),
  [REGISTRATION_ACTIONS.RESET_FORM_VALUES]: state =>
    Object.assign({}, state, {
      formValues: defaultRegistrationData.formValues,
    }),
});
