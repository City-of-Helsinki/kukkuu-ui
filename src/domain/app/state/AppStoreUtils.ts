import { persistor, store } from './AppStore';
import { resetBackendAuthentication } from '../../auth/state/BackendAuthenticationActions';
import client from '../../api/client';
import { clearProfile } from '../../profile/state/ProfileActions';
import { resetFormValues } from '../../registration/state/RegistrationActions';

/**
 * Flush entire Redux data store
 * as well as persisted data
 * Apollo cache
 */
export const flushData = () => {
  // Clear backend auth data
  store.dispatch(resetBackendAuthentication());
  // Clear user form data
  store.dispatch(resetFormValues());
  // Clear profile (fetched from API)
  store.dispatch(clearProfile());
  // Flush data in redux store and localStorage
  persistor.flush();
  // Clear Apollo cache
  client.clearStore();
};
