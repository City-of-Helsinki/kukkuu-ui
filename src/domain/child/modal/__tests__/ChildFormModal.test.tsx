import ChildFormModal from '../ChildFormModal';
import { defaultRegistrationData } from '../../../registration/state/RegistrationReducers';
import { customRender as render } from '../../../../common/test/customRender';

it('renders snapshot correctly', () => {
  const initialValues = Object.assign(
    {},
    defaultRegistrationData.formValues.children[0],
    { birthyear: 2019 }
  );

  const { container } = render(
    <ChildFormModal
      initialValues={initialValues}
      onSubmit={vi.fn()}
      label="foo"
      isOpen={false}
      setIsOpen={vi.fn()}
      onCancel={vi.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});
