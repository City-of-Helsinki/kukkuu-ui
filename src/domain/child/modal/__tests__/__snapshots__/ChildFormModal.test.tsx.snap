// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<div
  className="childFormModalWrapper"
>
  <Modal
    isOpen={false}
    label="foo"
    setFormIsFilling={[Function]}
    toggleModal={[Function]}
  >
    <ChildForm
      formType="ADD"
      initialValues={
        Object {
          "birthdate": "01-11-2019",
          "firstName": "",
          "homeCity": "",
          "lastName": "",
          "postalCode": "",
          "relationship": Object {
            "type": undefined,
          },
        }
      }
      onSubmit={[MockFunction]}
      setFormIsFilling={[Function]}
    />
  </Modal>
</div>
`;
